extensions/amp-autocomplete/0.1/amp-autocomplete.js

```diff
@@ -446,23 +544,65 @@ export class AmpAutocomplete extends AMP.BaseElement {
    * @private
    */
   inputHandler_() {
+    if (
+      !this.binding_.shouldAutocomplete(
+        this.trigger_,
+        dev().assertElement(this.inputElement_)
+      )
+    ) {
+      return this.mutateElement(() => {
+        this.clearAllItems_();
+      });
+    }
+
     // If the input is the first entry in the field.
-    const firstEntry =
+    const firstInteraction =
       this.userInput_.length === 0 && this.inputElement_.value.length === 1;
-    this.userInput_ = this.inputElement_.value;
-    return this.mutateElement(() => {
-      this.filterDataAndRenderResults_(this.sourceData_, this.userInput_);
-      this.toggleResults_(true);
-      if (this.suggestFirst_) {
-        if (!this.detectBackspace_ || firstEntry) {
-          this.updateActiveItem_(1);
-        }
-        // Reset flag.
-        this.detectBackspace_ = false;
-      }
+    this.userInput_ = this.binding_.updateUserInput(
+      this.trigger_,
+      dev().assertElement(this.inputElement_)
+    );
+
+    return this.maybeFetchAndAutocomplete_(firstInteraction);
+  }
+
+  /**
+   * @param {boolean} firstInteraction
+   * @return {!Promise}
+   */
+  maybeFetchAndAutocomplete_(firstInteraction) {
+    // Always fetch in email context, where data is required to be dynamic.
+    const fetchIfEmail = this.isEmail_
```

Comment:
Can this condition be contained in `this.getRemoteData_()` and always return a promise?

Reply:
I could be misunderstanding your suggestion, but seeing as `getRemoteData_` can be called under different conditions elsewhere, we wouldn't want it to always have the behavior of early-ing out or returning stale data (`this.sourceData_`) if it's not an email.

Reply:
For emails, we want to fetch on every input. For non-email, other behaviors (including `mutatedAttributesCallback` on the `src` attribute) can and should trigger `getRemoteData_`

Reply:
Yup, I missed the second usage. Fine to leave as-is but now I'm wondering about the signature for `maybeFetchAndAutocomplete_`.

It seems like you're only using it once so can we evaluate `isFirstInteraction` there to prevent drilling down a boolean?

Reply:
Good point. With that modification I ended up reversing the clauses anyway. 😛